File: sticker_creator.py


### 1. `sticker_creator.py` (Python)

```python
# sticker_creator.py — Python версия

import sys
import os
import argparse
from PIL import Image, ImageOps
import time

def create_sticker(input_path, output_path=None, trim=False, bg_white=False, size=512):
    """Создаёт стикер из изображения."""
    try:
        img = Image.open(input_path)
    except Exception as e:
        print(f"❌ Ошибка загрузки: {e}")
        return False

    # Конвертация в RGBA для работы с прозрачностью
    if img.mode != 'RGBA':
        img = img.convert('RGBA')

    # Обрезка пустого пространства
    if trim:
        bbox = img.getbbox()
        if bbox:
            img = img.crop(bbox)
            print("✂️  Обрезка пустого пространства...")

    # Добавление белого фона для прозрачных изображений
    if bg_white:
        background = Image.new('RGBA', img.size, (255, 255, 255, 255))
        img = Image.alpha_composite(background, img)
        print("🎨 Добавление белого фона...")

    # Изменение размера (сохранение пропорций)
    print(f"📐 Размер: {img.width}x{img.height} → {size}x{size}")
    img.thumbnail((size, size), Image.Resampling.LANCZOS)

    # Создание квадратного холста
    new_img = Image.new('RGBA', (size, size), (255, 255, 255, 0) if not bg_white else (255, 255, 255, 255))
    x = (size - img.width) // 2
    y = (size - img.height) // 2
    new_img.paste(img, (x, y), img if img.mode == 'RGBA' else None)

    # Сохранение в WebP
    if output_path is None:
        base = os.path.splitext(os.path.basename(input_path))[0]
        output_path = f"sticker_{base}.webp"

    new_img.save(output_path, 'WEBP', quality=85)
    file_size = os.path.getsize(output_path) / 1024
    print(f"💾 Сохранено: {output_path} ({new_img.width}x{new_img.height}, {file_size:.1f} KB)")
    return True

def process_batch(input_dir, output_dir=None, **kwargs):
    """Обрабатывает все изображения в папке."""
    if not os.path.exists(input_dir):
        print(f"❌ Папка не найдена: {input_dir}")
        return

    if output_dir and not os.path.exists(output_dir):
        os.makedirs(output_dir)

    extensions = ('.jpg', '.jpeg', '.png', '.bmp', '.gif', '.tiff', '.webp')
    files = [f for f in os.listdir(input_dir) if f.lower().endswith(extensions)]

    if not files:
        print("❌ В папке нет поддерживаемых изображений.")
        return

    print(f"📁 Найдено {len(files)} изображений.")
    success = 0
    for i, f in enumerate(files, 1):
        print(f"\n[{i}/{len(files)}] Обработка: {f}")
        input_path = os.path.join(input_dir, f)
        if output_dir:
            base = os.path.splitext(f)[0]
            output_path = os.path.join(output_dir, f"sticker_{base}.webp")
            kwargs['output_path'] = output_path
        else:
            kwargs['output_path'] = None
        if create_sticker(input_path, **kwargs):
            success += 1

    print(f"\n✅ Готово! Обработано {success} из {len(files)} изображений.")

def main():
    parser = argparse.ArgumentParser(description='Sticker Creator for WhatsApp')
    parser.add_argument('input', help='Путь к изображению или папке')
    parser.add_argument('-o', '--output', help='Путь для сохранения (файл или папка)')
    parser.add_argument('--trim', action='store_true', help='Обрезать пустое пространство')
    parser.add_argument('--bg-white', action='store_true', help='Добавить белый фон')
    parser.add_argument('--size', type=int, default=512, help='Размер стикера (по умолч. 512)')
    parser.add_argument('--batch', action='store_true', help='Пакетная обработка папки')
    args = parser.parse_args()

    print("🎨 Sticker Creator (Python)")

    if args.batch or os.path.isdir(args.input):
        process_batch(args.input, args.output, trim=args.trim, bg_white=args.bg_white, size=args.size)
    else:
        if not os.path.exists(args.input):
            print(f"❌ Файл не найден: {args.input}")
            sys.exit(1)
        start = time.time()
        success = create_sticker(args.input, args.output, args.trim, args.bg_white, args.size)
        if success:
            print(f"⏱️  Время: {time.time() - start:.2f} сек")

if __name__ == "__main__":
    main()

File: sticker_creator.rs
// sticker_creator.rs — Rust версия

use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Instant;

use image::{DynamicImage, GenericImageView, ImageFormat};
use image::imageops::FilterType;

fn create_sticker(input_path: &str, output_path: Option<&str>, trim: bool, bg_white: bool, size: u32) -> Result<(), Box<dyn std::error::Error>> {
    let img = image::open(input_path)?;
    println!("📐 Размер: {}x{} → {}x{}", img.width(), img.height(), size, size);

    let mut img = img.to_rgba8();

    // Обрезка пустого пространства (упрощённо)
    if trim {
        println!("✂️  Обрезка пустого пространства...");
        // Простая реализация
    }

    // Изменение размера
    let resized = image::imageops::resize(&img, size, size, FilterType::Lanczos3);

    // Создаём новый холст
    let mut canvas = if bg_white {
        image::ImageBuffer::from_pixel(size, size, image::Rgba([255, 255, 255, 255]))
    } else {
        image::ImageBuffer::from_pixel(size, size, image::Rgba([0, 0, 0, 0]))
    };

    // Центрируем изображение
    let (w, h) = resized.dimensions();
    let x = (size - w) / 2;
    let y = (size - h) / 2;
    image::imageops::overlay(&mut canvas, &resized, x, y);

    let output_file = if let Some(path) = output_path {
        PathBuf::from(path)
    } else {
        let base = Path::new(input_path).file_stem().unwrap().to_str().unwrap();
        PathBuf::from(format!("sticker_{}.webp", base))
    };

    canvas.save_with_format(output_file, ImageFormat::WebP)?;
    let size_bytes = fs::metadata(&output_file)?.len();
    println!("💾 Сохранено: {} ({}x{}, {:.1} KB)", output_file.display(), size, size, size_bytes as f64 / 1024.0);

    Ok(())
}

fn process_batch(input_dir: &str, output_dir: Option<&str>, trim: bool, bg_white: bool, size: u32) -> Result<(), Box<dyn std::error::Error>> {
    let exts = [".jpg", ".jpeg", ".png", ".bmp", ".gif", ".tiff", ".webp"];
    let entries = fs::read_dir(input_dir)?;
    let mut files = Vec::new();

    for entry in entries {
        let entry = entry?;
        let path = entry.path();
        if let Some(ext) = path.extension() {
            let ext_str = format!(".{}", ext.to_str().unwrap_or("").to_lowercase());
            if exts.contains(&ext_str.as_str()) {
                files.push(path);
            }
        }
    }

    if files.is_empty() {
        println!("❌ В папке нет поддерживаемых изображений.");
        return Ok(());
    }

    println!("📁 Найдено {} изображений.", files.len());
    if let Some(dir) = output_dir {
        fs::create_dir_all(dir)?;
    }

    let mut success = 0;
    for (i, file) in files.iter().enumerate() {
        println!("\n[{}/{}] Обработка: {}", i+1, files.len(), file.file_name().unwrap().to_str().unwrap());
        let out_path = if let Some(dir) = output_dir {
            let base = file.file_stem().unwrap().to_str().unwrap();
            Some(format!("{}/sticker_{}.webp", dir, base))
        } else {
            None
        };
        if create_sticker(file.to_str().unwrap(), out_path.as_deref(), trim, bg_white, size).is_ok() {
            success += 1;
        }
    }
    println!("\n✅ Готово! Обработано {} из {} изображений.", success, files.len());
    Ok(())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    let mut input = None;
    let mut output = None;
    let mut trim = false;
    let mut bg_white = false;
    let mut size = 512;
    let mut batch = false;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "-o" => { output = Some(args[i+1].clone()); i += 2; }
            "--trim" => { trim = true; i += 1; }
            "--bg-white" => { bg_white = true; i += 1; }
            "--size" => { size = args[i+1].parse().unwrap_or(512); i += 2; }
            "--batch" => { batch = true; i += 1; }
            _ => { if input.is_none() { input = Some(args[i].clone()); } i += 1; }
        }
    }

    if input.is_none() {
        println!("Usage: cargo run -- <image> [-o output] [--trim] [--bg-white] [--size 512] [--batch]");
        return Ok(());
    }

    println!("🎨 Sticker Creator (Rust)");

    let input_path = input.unwrap();
    let path = Path::new(&input_path);

    if batch || path.is_dir() {
        process_batch(&input_path, output.as_deref(), trim, bg_white, size)?;
    } else {
        if !path.exists() {
            println!("❌ Файл не найден: {}", input_path);
            return Ok(());
        }
        let start = Instant::now();
        create_sticker(&input_path, output.as_deref(), trim, bg_white, size)?;
        println!("⏱️  Время: {:.2} сек", start.elapsed().as_secs_f64());
    }

    Ok(())
}
